File: source/com/barrybecker4/simulation/conway/model/ConwayProcessor.java
package com.barrybecker4.simulation.conway.model;

import com.barrybecker4.common.geometry.IntLocation;
import com.barrybecker4.common.geometry.Location;
import com.barrybecker4.simulation.conway.model.rules.*;

import java.util.Set;

/**
 * @author Barry Becker
 */
public class ConwayProcessor {

    /** cells die if less than this */
    public static final boolean DEFAULT_USE_PARALLEL = true;
    private boolean wrapGrid = false;
    private int width = -1;
    private int height = -1;
    private Rule rule = new RuleB3S23(); // new RuleB34S456Amoeba();

    public enum RuleType {TraditionalB3S23, AmoebaB34S456, HighlifeB36S23, SwarmB356S23}
    public static final RuleType DEFAULT_RULE_TYPE = RuleType.TraditionalB3S23;

    private Conway conway;
    /** Manages the worker threads. */
    //private RunnableParallelizer parallelizer;

    /** Constructor that allows you to specify the dimensions of the conway */
    public ConwayProcessor() {
        this(DEFAULT_USE_PARALLEL);
    }

    ConwayProcessor(boolean useParallel) {
        conway = new Conway();
        conway.initialize();
        setUseParallel(useParallel);
    }

    void setWrap(boolean wrapGrid, int width, int height) {
        this.wrapGrid = wrapGrid;
        this.width = width;
        this.height = height;
    }

    void setRuleType(RuleType type) {
        switch (type) {
            case TraditionalB3S23: rule = new RuleB3S23(); break;
            case AmoebaB34S456: rule = new RuleB34S456Amoeba(); break;
            case HighlifeB36S23: rule = new RuleB36S23Highlife(); break;
            case SwarmB356S23: rule = new RuleB356S23Swarm(); break;
        }
    }

    void setUseParallel(boolean parallelized) {
        //parallelizer =
        //     parallelized ? new RunnableParallelizer() : new RunnableParallelizer(1);
    }

    public Set<Location> getPoints() {
        return conway.getPoints();
    }

    void setAlive(int row, int col) {
        conway.setValue(new IntLocation(row, col), 1);
    }

    /**
     * Compute the next step of the simulation.
     */
    void nextPhase() {
        Conway newConway = new Conway();
        newConway.setWrapping(wrapGrid, width, height);

        conway = rule.applyRule(conway, newConway);

        /*
        int numThreads = parallelizer.getNumThreads();
        List<Runnable> workers = new ArrayList<>(numThreads + 1);
        int range = conway.getNumPoints() / numThreads;
        for (int i = 0; i < numThreads; i++) {
            int offset = i * range;
            workers.add(new Worker(offset, offset + range, newConway));
        }

        // blocks until all Callables are done running.
        parallelizer.invokeAllRunnables(workers);
        */
    }

    public Integer getValue (Location c) {
        return conway.getValue(c);
    }

    public String toString() {
        return conway.toString();
    }


    public static void main(String[] args) {
        ConwayProcessor cave = new ConwayProcessor(DEFAULT_USE_PARALLEL);
        cave.nextPhase();
    }
}

File: scala-source/com/barrybecker4/simulation/conway/model/ConwayProcessor.scala
// Copyright by Barry G. Becker, 2014-2017. Licensed under MIT License: http://www.opensource.org/licenses/MIT
package com.barrybecker4.simulation.conway.model

import com.barrybecker4.common.geometry.IntLocation
import com.barrybecker4.common.geometry.Location
import com.barrybecker4.simulation.conway.model.rules.{Rule, TotalisticLifeRule}
import ConwayProcessor.RuleType
import ConwayProcessor.RuleType.{AmoebaB34S456, HighlifeB36S23, SwarmB356S23, TraditionalB3S23}

/**
  * @author Barry Becker
  */
object ConwayProcessor {

  val DEFAULT_USE_PARALLEL = false

  enum RuleType:
    case TraditionalB3S23, AmoebaB34S456, HighlifeB36S23, SwarmB356S23

  val DEFAULT_RULE_TYPE: RuleType = RuleType.TraditionalB3S23
}

class ConwayProcessor private[model](var useParallel: Boolean = ConwayProcessor.DEFAULT_USE_PARALLEL) {

  private var conway = new Conway
  private var wrapGrid = false
  private var width = -1
  private var height = -1
  private var rule: Rule = TotalisticLifeRule(Set(3), Set(2, 3))
  conway.initialize()

  private[model] def setWrap(wrapGrid: Boolean, width: Int, height: Int): Unit = {
    this.wrapGrid = wrapGrid
    this.width = width
    this.height = height
  }

  private[model] def setRuleType(ruleType: RuleType): Unit = {
    rule = ruleType match {
      case TraditionalB3S23 => TotalisticLifeRule(Set(3), Set(2, 3))
      case AmoebaB34S456 => TotalisticLifeRule(Set(3, 4), Set(4, 5, 6))
      case HighlifeB36S23 => TotalisticLifeRule(Set(3, 6), Set(2, 3))
      case SwarmB356S23 => TotalisticLifeRule(Set(3, 5, 6), Set(2, 3))
    }
  }

  def getPoints: Set[Location] = conway.getPoints

  private[model] def setAlive(row: Int, col: Int): Unit =
    conway.setValue(IntLocation(row, col), 1)

  private[model] def setDead(row: Int, col: Int): Unit =
    conway.setValue(IntLocation(row, col), 0)

  private[model] def clearGridForTesting(): Unit =
    conway.clearForTesting()

  /** Compute the next step of the simulation. */
  private[model] def nextPhase(): Unit = {
    val newConway = new Conway
    newConway.setWrapping(wrapGrid, width, height)
    conway = rule.applyRule(conway, newConway, useParallel)
  }

  def getValue(c: Location): Int = conway.getValue(c)

  def isAlive(c: Location): Boolean = conway.isAlive(c)

  override def toString: String = conway.toString
}
